cmd/calico-ipam-upgrade/main.go

```diff
@@ -0,0 +1,193 @@
+// Copyright 2015 Tigera Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package main
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/projectcalico/libcalico-go/lib/apiconfig"
+	client "github.com/projectcalico/libcalico-go/lib/clientv3"
+	"github.com/projectcalico/libcalico-go/lib/ipam"
+	"github.com/projectcalico/libcalico-go/lib/logutils"
+	cnet "github.com/projectcalico/libcalico-go/lib/net"
+)
+
+const (
+	ipAllocPath = "/var/lib/cni/networks/k8s-pod-network"
+)
+
+var (
+	binariesToDisable = []string{
+		"/host/opt/cni/bin/calico",
+		"/host/opt/cni/bin/host-local",
+	}
+)
+
+func main() {
+	// Set up logging formatting.
+	log.SetFormatter(&logutils.Formatter{})
+
+	// Install a hook that adds file/line no information.
+	log.AddHook(&logutils.ContextHook{})
+
+	// Check to see if the system is still using host-local
+	// by checking the existence of the path.
+	var err error
+	if _, err = os.Stat(ipAllocPath); err != nil && os.IsNotExist(err) {
+		log.Info("no migration necessary, exiting...")
+		os.Exit(0)
+	}
+
+	// First, disable the necessary binaries so it can't be used.
+	for _, binary := range binariesToDisable {
+		if err = toggleFile(binary, true); err != nil {
+			log.WithField("file", binary).WithError(err).Fatal("failed to disable binary")
+		}
+	}
+	defer func() {
+		if err != nil {
+			rollback()
+		}
+	}()
+	<-time.After(time.Second)
+
+	// Initialize necessary variables
+	nodename := os.Getenv("KUBERNETES_NODE_NAME")
+	cfg, err := apiconfig.LoadClientConfig("")
+	if err != nil {
+		log.WithError(err).Fatal("failed to load api client config")
+	}
+	cfg.Spec.DatastoreType = apiconfig.Kubernetes
+	calicoClient, err := client.New(*cfg)
+	if err != nil {
+		log.WithError(err).Fatal("failed to initialize api client")
+	}
+
+	// Disable auto-allocation of blocks
+	if err = calicoClient.IPAM().SetIPAMConfig(context.Background(), ipam.IPAMConfig{
```

Comment:
Let's swap this over to using the ready flag for now - I think given these lines of code this won't work the way I originally imagined: https://github.com/projectcalico/libcalico-go/blob/9ccf86fcbbe260dd6ba0c31afaa9686e3025868e/lib/ipam/ipam.go#L1357-L1360